packages/expo-camera/android/src/main/java/expo/modules/camera/next/tasks/ResolveTakePicture.kt

```diff
@@ -0,0 +1,248 @@
+private const val DIRECTORY_NAME = "Camera"
+private const val EXTENSION = ".jpg"
+private const val BASE64_KEY = "base64"
+private const val HEIGHT_KEY = "height"
+private const val WIDTH_KEY = "width"
+private const val EXIF_KEY = "exif"
+private const val DATA_KEY = "data"
+private const val URI_KEY = "uri"
+private const val ID_KEY = "id"
+
+class ResolveTakePicture(
+  private var imageData: ByteArray,
+  private var promise: Promise,
+  private var options: PictureOptions,
+  private val directory: File,
+  private var pictureSavedDelegate: PictureSavedDelegate
+) {
+
+  private val quality: Int
+    get() = (options.quality * 100).toInt()
+
+  suspend fun resolve() = withContext(Dispatchers.IO) {
+    val bundle = processImage()
+    onComplete(bundle)
+  }
+
+  private fun processImage(): Bundle? {
+    // handle SkipProcessing
+    if (options.skipProcessing) {
+      return skipProcessing()
+    }
+
+    // set, read, and apply EXIF data
+    try {
+      ByteArrayInputStream(imageData).use { inputStream ->
+        val response = Bundle()
+
+        val exifInterface = ExifInterface(inputStream)
+
+        // If there are additional exif data, insert it here
+        options.additionalExif?.let {
+          setExifData(exifInterface, it)
+        }
+
+        // Get orientation of the image from mImageData via inputStream
+        val orientation = exifInterface.getAttributeInt(
+          ExifInterface.TAG_ORIENTATION,
+          ExifInterface.ORIENTATION_UNDEFINED
+        )
+
+        val bitmapOptions = BitmapFactory
+          .Options()
+          .apply {
+            inSampleSize = 1
+          }
+        var bitmap: Bitmap? = null
+        var lastError: Error? = null
+
+        // If OOM exception was thrown, we try to use downsampling to recover.
+        while (bitmapOptions.inSampleSize <= options.maxDownsampling) {
+          try {
+            bitmap = decodeBitmap(imageData, orientation, bitmapOptions)
+            break
+          } catch (exception: OutOfMemoryError) {
+            bitmapOptions.inSampleSize *= 2
+            lastError = exception
+          }
+        }
+
+        if (bitmap == null) {
+          promise.reject(OUT_OF_MEMORY_TAG, OUT_OF_MEMORY_EXCEPTION_MSG, lastError)
+          return null
+        }
+
+        // Write Exif data to the response if requested
+        if (options.exif) {
+          val exifData = getExifData(exifInterface)
+          response.putBundle(EXIF_KEY, exifData)
+        }
+
+        // Upon rotating, write the image's dimensions to the response
+        response.apply {
+          putInt(WIDTH_KEY, bitmap.width)
+          putInt(HEIGHT_KEY, bitmap.height)
+        }
+
+        // Cache compressed image in imageStream
+        ByteArrayOutputStream().use { imageStream ->
+          bitmap.compress(Bitmap.CompressFormat.JPEG, quality, imageStream)
+          // Write compressed image to file in cache directory
+          val filePath = writeStreamToFile(imageStream)
```

Comment:
nit: maybe to `bitmap.recycle()` to free unused memory earlier.
or we can also use the kotlin [use](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.io/use.html) statement